### PR TITLE
[DefectId: 39276]Removed Target Application that exist in businessflow but not in activities of the business flow

### PR DESCRIPTION
**Pull request by manas-droid**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [ ] PR description and commit message should describe the changes done in this PR
- [ ] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [ ] Latest Code from master or specific release branch is merged to your branch
- [ ] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

### CODE REVIEW

File: Ginger/GingerCoreCommon/Repository/BusinessFlowLib/BusinessFlow.cs [hunk 1]
```diff
@@ -345,6 +345,20 @@ private void DoActivitiesLazyLoad()
                 {
                     this.TrackObservableList(mActivities);
                 }
+
+                IEnumerable<string> distinctTargetApp =  mActivities.Select((activity) => activity.TargetApplication).Distinct();
+
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/Repository/BusinessFlowLib/BusinessFlow.cs:350]: If now the Activities selected Target Applications set the Business Flow list of Target Applications so no need to maintain “TargetApplications” list on BF level which will be serialized, we can dynamically each time return the current selected Applications on all child Activities.

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 2]
```diff
@@ -646,22 +647,22 @@ private void OnTargetApplicationChanged(object arg1, SelectionChangedEventArgs a
...
                     // If not present, add to the removal list
-                    agentsToRemove.Add(existingTargetApp);
+                    TargetApplicationsToRemove.Add(existingTargetApp);
                 }
             }
 
-            // Remove the agents from mExecutionEngine.GingerRunner.ApplicationAgents
-            foreach (var agentToRemove in agentsToRemove)
+            // Remove the target applications from mBusinessFlow.TargetApplications
+            foreach (var agentToRemove in TargetApplicationsToRemove)
```
Comment (MeniKadosh1) [inline Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:665]: you missed the "agentToRemove"

File: Ginger/GingerCoreNET/Run/GingerExecutionEngine.cs [hunk 3]
```diff
@@ -49,6 +49,7 @@
 using GingerWPF.GeneralLib;
 using Microsoft.CodeAnalysis.VisualBasic.Syntax;
 using System;
+using System.Collections;
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/Run/GingerExecutionEngine.cs:52]: this was added for. a reason?

File: Ginger/GingerCoreCommon/Repository/BusinessFlowLib/BusinessFlow.cs (hunk 1 above)
Comment (MeniKadosh1) [reply Ginger/GingerCoreCommon/Repository/BusinessFlowLib/BusinessFlow.cs:350]: seems like there are many calls to TargetApplicaitons field and also the Activities having only mapping by name so relaying only on Activities can be a problem.

File: Ginger/GingerCoreNET/Run/GingerExecutionEngine.cs (hunk 3 above)
Comment (manas-droid) [reply Ginger/GingerCoreNET/Run/GingerExecutionEngine.cs:52]: Added this accidentally , will remove this and other unused packages

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs (hunk 2 above)
Comment (manas-droid) [reply Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs]: Implemented

**Review by MeniKadosh1 — APPROVED** [review]

**Review by MeniKadosh1 — APPROVED** [review]

**Review by MeniKadosh1 — APPROVED** [review]

**Review by MeniKadosh1 — APPROVED** [review]

**Pull request merged**